modules/grpc/bigquery/bigquery-worker.cpp

```diff
@@ -0,0 +1,516 @@
+
+  this->prepare_batch();
+
+  msg_debug("Connecting to BigQuery", log_pipe_location_tag((LogPipe *) this->super->super.owner));
+
+  std::chrono::system_clock::time_point connect_timeout =
+    std::chrono::system_clock::now() + std::chrono::seconds(10);
+
+  if (!this->channel->WaitForConnected(connect_timeout))
+    return false;
+
+  this->connected = true;
+  return true;
+}
+
+void
+DestinationWorker::disconnect()
+{
+  if (!this->connected)
+    return;
+
+  if (!this->batch_writer->WritesDone())
+    msg_warning("Error closing BigQuery write stream, writes may have been unsuccessful",
+                log_pipe_location_tag((LogPipe *) this->super->super.owner));
+
+  ::grpc::Status status = this->batch_writer->Finish();
+  if (!status.ok() && status.error_code() != ::grpc::StatusCode::INVALID_ARGUMENT)
+    {
+      msg_warning("Error closing BigQuery stream", evt_tag_str("error", status.error_message().c_str()),
+                  evt_tag_str("details", status.error_details().c_str()),
+                  evt_tag_int("code", status.error_code()),
+                  log_pipe_location_tag((LogPipe *) this->super->super.owner));
+    }
+
+  ::grpc::ClientContext ctx;
+  google::cloud::bigquery::storage::v1::FinalizeWriteStreamRequest finalize_request;
+  google::cloud::bigquery::storage::v1::FinalizeWriteStreamResponse finalize_response;
+  finalize_request.set_name(write_stream.name());
+
+  status = this->stub->FinalizeWriteStream(&ctx, finalize_request, &finalize_response);
+  if (!status.ok())
+    {
+      msg_warning("Error finalizing BigQuery write stream", evt_tag_str("error", status.error_message().c_str()),
+                  evt_tag_str("details", status.error_details().c_str()),
+                  log_pipe_location_tag((LogPipe *) this->super->super.owner));
+    }
+
+  this->connected = false;
+}
+
+void
+DestinationWorker::prepare_batch()
+{
+  this->batch_size = 0;
+  this->current_batch = google::cloud::bigquery::storage::v1::AppendRowsRequest{};
+
+  this->current_batch.set_write_stream(write_stream.name());
+  this->current_batch.set_trace_id("syslog-ng-bigquery");
+  google::cloud::bigquery::storage::v1::AppendRowsRequest_ProtoData *proto_rows =
+    this->current_batch.mutable_proto_rows();
+  google::cloud::bigquery::storage::v1::ProtoSchema *schema = proto_rows->mutable_writer_schema();
+  this->get_owner()->schema_descriptor->CopyTo(schema->mutable_proto_descriptor());
+}
+
+void
+DestinationWorker::format_template(LogTemplate *tmpl, LogMessage *msg, GString *value, LogMessageValueType *type)
+{
+  DestinationDriver *owner = this->get_owner();
+
+  if (log_template_is_trivial(tmpl))
+    {
+      gssize trivial_value_len;
+      const gchar *trivial_value = log_template_get_trivial_value_and_type(tmpl, msg, &trivial_value_len, type);
+
+      g_string_truncate(value, 0);
+      if (trivial_value)
+        g_string_append_len(value, trivial_value, trivial_value_len);
+
+      return;
+    }
```

Comment:
I think we could skip this string copy if the `format_template()` function returned a `const char *`, which in case of a trivial value would point to the result of `log_template_get_trivial_value_and_type()` or in case of a non-trivial value it would point to `value->str`.

The call site could use the returned `const char*`, which either points to the scratch buffer's value or to a trivial value, it does not care.

Reply:
Thanks, I will do it soon. :)

Reply:
Hmm, it turned out I can't do this, because `TYPE_BYTES` allows embedded NULLs and does not assume NULL-termination. This is covered by the `CPPTYPE_STRING` case:

```
 CPPTYPE_STRING = 9,    // TYPE_STRING, TYPE_BYTES
```

Reply:
I think it could still work:

`log_template_get_trivial_value_and_type()` returns the length, too, and the `format_template()` function could return the `const char *` and the trivial values length, or the scratch buffer's `str` and `len`.

Reply:
I know, but does it worth it, especially in C++?

Reply:
Sorry, I don't understand your last question. It's okay to pass it to the C++ function as an `std::string`, of course.
What I propose is skipping the unnecessary copy of the trivial value to the Scratch Buffer GString, which I think is worth to do.

Somethink like this:
```diff
diff --git a/modules/grpc/bigquery/bigquery-worker.cpp b/modules/grpc/bigquery/bigquery-worker.cpp
index 122edc0a4..15694525f 100644
--- a/modules/grpc/bigquery/bigquery-worker.cpp
+++ b/modules/grpc/bigquery/bigquery-worker.cpp
@@ -176,25 +176,22 @@ DestinationWorker::prepare_batch()
   this->get_owner()->schema_descriptor->CopyTo(schema->mutable_proto_descriptor());
 }
 
-void
-DestinationWorker::format_template(LogTemplate *tmpl, LogMessage *msg, GString *value, LogMessageValueType *type)
+const char *
+DestinationWorker::format_template(LogTemplate *tmpl, LogMessage *msg, GString *buffer, LogMessageValueType *type,
+                                   gssize *len)
 {
   DestinationDriver *owner = this->get_owner();
 
   if (log_template_is_trivial(tmpl))
     {
-      gssize trivial_value_len;
-      const gchar *trivial_value = log_template_get_trivial_value_and_type(tmpl, msg, &trivial_value_len, type);
-
-      g_string_truncate(value, 0);
-      if (trivial_value)
-        g_string_append_len(value, trivial_value, trivial_value_len);
-
-      return;
+      const gchar *trivial_value = log_template_get_trivial_value_and_type(tmpl, msg, len, type);
+      return trivial_value;
     }
 
   LogTemplateEvalOptions options = {&owner->template_options, LTZ_SEND, this->super->super.seq_num, NULL, LM_VT_STRING};
-  log_template_format_value_and_type(tmpl, msg, &options, value, type);
+  log_template_format_value_and_type(tmpl, msg, &options, buffer, type);
+  *len = buffer->len;
+  return buffer->str;
 }
 
 bool
@@ -204,11 +201,12 @@ DestinationWorker::insert_field(const google::protobuf::Reflection *reflection,
   DestinationDriver *owner = this->get_owner();
 
   ScratchBuffersMarker m;
-  GString *value = scratch_buffers_alloc_and_mark(&m);
+  GString *buffer = scratch_buffers_alloc_and_mark(&m);
 
   LogMessageValueType type;
 
-  this->format_template(field.value, msg, value, &type);
+  gssize len;
+  const char *value = this->format_template(field.value, msg, buffer, &type, &len);
 
   if (type == LM_VT_NULL)
     {
@@ -225,14 +223,14 @@ DestinationWorker::insert_field(const google::protobuf::Reflection *reflection,
   switch (field.field_desc->cpp_type())
     {
     case FieldDescriptor::CppType::CPPTYPE_STRING:
-      reflection->SetString(message, field.field_desc, std::string{value->str, value->len});
+      reflection->SetString(message, field.field_desc, std::string{value, (gsize) len});
       break;
     case FieldDescriptor::CppType::CPPTYPE_INT32:
     {
       int32_t v;
-      if (!type_cast_to_int32(value->str, &v, NULL))
+      if (!type_cast_to_int32(value, &v, NULL))
         {
-          type_cast_drop_helper(owner->template_options.on_error, value->str, "integer");
+          type_cast_drop_helper(owner->template_options.on_error, value, "integer");
           goto error;
         }
       reflection->SetInt32(message, field.field_desc, v);
@@ -241,9 +239,9 @@ DestinationWorker::insert_field(const google::protobuf::Reflection *reflection,
     case FieldDescriptor::CppType::CPPTYPE_INT64:
     {
       int64_t v;
-      if (!type_cast_to_int64(value->str, &v, NULL))
+      if (!type_cast_to_int64(value, &v, NULL))
         {
-          type_cast_drop_helper(owner->template_options.on_error, value->str, "integer");
+          type_cast_drop_helper(owner->template_options.on_error, value, "integer");
           goto error;
         }
       reflection->SetInt64(message, field.field_desc, v);
@@ -252,9 +250,9 @@ DestinationWorker::insert_field(const google::protobuf::Reflection *reflection,
     case FieldDescriptor::CppType::CPPTYPE_UINT32:
     {
       int64_t v;
-      if (!type_cast_to_int64(value->str, &v, NULL))
+      if (!type_cast_to_int64(value, &v, NULL))
         {
-          type_cast_drop_helper(owner->template_options.on_error, value->str, "integer");
+          type_cast_drop_helper(owner->template_options.on_error, value, "integer");
           goto error;
         }
       reflection->SetUInt32(message, field.field_desc, (uint32_t) v);
@@ -263,9 +261,9 @@ DestinationWorker::insert_field(const google::protobuf::Reflection *reflection,
     case FieldDescriptor::CppType::CPPTYPE_UINT64:
     {
       int64_t v;
-      if (!type_cast_to_int64(value->str, &v, NULL))
+      if (!type_cast_to_int64(value, &v, NULL))
         {
-          type_cast_drop_helper(owner->template_options.on_error, value->str, "integer");
+          type_cast_drop_helper(owner->template_options.on_error, value, "integer");
           goto error;
         }
       reflection->SetUInt64(message, field.field_desc, (uint64_t) v);
@@ -274,9 +272,9 @@ DestinationWorker::insert_field(const google::protobuf::Reflection *reflection,
     case FieldDescriptor::CppType::CPPTYPE_DOUBLE:
     {
       double v;
-      if (!type_cast_to_double(value->str, &v, NULL))
+      if (!type_cast_to_double(value, &v, NULL))
         {
-          type_cast_drop_helper(owner->template_options.on_error, value->str, "double");
+          type_cast_drop_helper(owner->template_options.on_error, value, "double");
           goto error;
         }
       reflection->SetDouble(message, field.field_desc, v);
@@ -285,9 +283,9 @@ DestinationWorker::insert_field(const google::protobuf::Reflection *reflection,
     case FieldDescriptor::CppType::CPPTYPE_FLOAT:
     {
       double v;
-      if (!type_cast_to_double(value->str, &v, NULL))
+      if (!type_cast_to_double(value, &v, NULL))
         {
-          type_cast_drop_helper(owner->template_options.on_error, value->str, "double");
+          type_cast_drop_helper(owner->template_options.on_error, value, "double");
           goto error;
         }
       reflection->SetFloat(message, field.field_desc, (float) v);
@@ -296,9 +294,9 @@ DestinationWorker::insert_field(const google::protobuf::Reflection *reflection,
     case FieldDescriptor::CppType::CPPTYPE_BOOL:
     {
       gboolean v;
-      if (!type_cast_to_boolean(value->str, &v, NULL))
+      if (!type_cast_to_boolean(value, &v, NULL))
         {
-          type_cast_drop_helper(owner->template_options.on_error, value->str, "boolean");
+          type_cast_drop_helper(owner->template_options.on_error, value, "boolean");
           goto error;
         }
       reflection->SetBool(message, field.field_desc, v);
diff --git a/modules/grpc/bigquery/bigquery-worker.hpp b/modules/grpc/bigquery/bigquery-worker.hpp
index 5fc4c64b2..61cb7cbec 100644
--- a/modules/grpc/bigquery/bigquery-worker.hpp
+++ b/modules/grpc/bigquery/bigquery-worker.hpp
@@ -63,7 +63,8 @@ private:
   bool insert_field(const google::protobuf::Reflection *reflection, const Field &field,
                     LogMessage *msg, google::protobuf::Message *message);
   LogThreadedResult handle_row_errors(const google::cloud::bigquery::storage::v1::AppendRowsResponse &response);
-  void format_template(LogTemplate *tmpl, LogMessage *msg, GString *value, LogMessageValueType *type);
+  const char *format_template(LogTemplate *tmpl, LogMessage *msg, GString *value, LogMessageValueType *type,
+                              gssize *len);
   DestinationDriver *get_owner();
 
 private:
```

Reply:
This is far away from C++ in my opinion. I've added "Slice" struct, trying to make it a little better.

Reply:
Thank you! Without this, I think we would lose all the benefits we gain from using `log_template_get_trivial_value_and_type()` .